File: lib/primitives.cc
#include "primitives.h"
#include "error.h"
#include "primitives_utils.h"

using namespace net_deserializer;

namespace
{
    enum class PrimitiveType : uint8_t
    {
        Boolean  = 1,
        Byte     = 2,
        Char     = 3,
        Decimal  = 5,
        Double   = 6,
        Int16    = 7,
        Int32    = 8,
        Int64    = 9,
        SByte    = 10,
        Single   = 11,
        TimeSpan = 12,
        DateTime = 13,
        UInt16   = 14,
        UInt32   = 15,
        UInt64   = 16,
        Null     = 17,
        String   = 18,
    };
}

std::unique_ptr<Node> net_deserializer::read_primitive(BinaryReader &reader)
{
    const auto primitive_type = reader.read<PrimitiveType>();

    if (primitive_type == PrimitiveType::Boolean)
        return make_primitive<bool>(reader);

    if (primitive_type == PrimitiveType::Byte)
        return make_primitive<uint8_t>(reader);

    if (primitive_type == PrimitiveType::Char)
        return make_primitive<char>(reader);

    if (primitive_type == PrimitiveType::Decimal)
        return make_primitive<float>(reader); // TODO: verify

    if (primitive_type == PrimitiveType::Double)
        return make_primitive<double>(reader);

    if (primitive_type == PrimitiveType::Int16)
        return make_primitive<int16_t>(reader);

    if (primitive_type == PrimitiveType::Int32)
        return make_primitive<int32_t>(reader);

    if (primitive_type == PrimitiveType::Int64)
        return make_primitive<int64_t>(reader);

    if (primitive_type == PrimitiveType::UInt16)
        return make_primitive<uint16_t>(reader);

    if (primitive_type == PrimitiveType::UInt32)
        return make_primitive<uint32_t>(reader);

    if (primitive_type == PrimitiveType::UInt64)
        return make_primitive<uint64_t>(reader);

    if (primitive_type == PrimitiveType::SByte)
        return make_primitive<int8_t>(reader);

    if (primitive_type == PrimitiveType::Single)
        return make_primitive<float>(reader);

    if (primitive_type == PrimitiveType::TimeSpan)
        return make_primitive<uint64_t>(reader); // TODO: verify

    if (primitive_type == PrimitiveType::DateTime)
        return make_primitive<uint64_t>(reader); // TODO: verify

    if (primitive_type == PrimitiveType::Null)
        return std::make_unique<PrimitiveNode>();

    if (primitive_type == PrimitiveType::String)
    {
        std::size_t length = 0;
        for (std::size_t i = 0; i < 5; i++)
        {
            const auto c = reader.read<uint8_t>();
            length <<= 7;
            length |= (c & 0x7F);
            if (!(c & 0x80))
                break;
        }
        const auto buf = std::make_unique<char[]>(length);
        for (std::size_t i = 0; i < length; i++)
            buf[i] = reader.read<char>();
        return std::make_unique<PrimitiveNode>(std::string(buf.get(), length));
    }

    throw NotImplementedError("Unknown primitive type: "
        + std::to_string(static_cast<int>(primitive_type)));
}

File: net_deserializer/primitives.py
from enum import Enum
from net_deserializer import util

class PrimitiveType(Enum):
    # pylint: disable=bad-whitespace
    Invalid  = 0
    Boolean  = 1
    Byte     = 2
    Char     = 3
    Decimal  = 5
    Double   = 6
    Int16    = 7
    Int32    = 8
    Int64    = 9
    SByte    = 10
    Single   = 11
    TimeSpan = 12
    DateTime = 13
    UInt16   = 14
    UInt32   = 15
    UInt64   = 16
    Null     = 17
    String   = 18

def _read_timespan(handle):
    return util.read_fmt(handle, '<q')

def _read_string(handle):
    length = 0
    for i in range(5):
        num = util.read_fmt(handle, 'B')
        length = length | (num & 0x7F) << (i * 7)
        if num & 0x80 == 0:
            break
    ret = handle.read(length)
    return ret.decode('utf-8')

_PRIMITIVE_FACTORY_MAP = {
    PrimitiveType.Boolean:  lambda h: util.read_fmt(h, '?'),
    PrimitiveType.Byte:     lambda h: util.read_fmt(h, 'B'),
    PrimitiveType.Double:   lambda h: util.read_fmt(h, 'd'),
    PrimitiveType.Int16:    lambda h: util.read_fmt(h, '<h'),
    PrimitiveType.Int32:    lambda h: util.read_fmt(h, '<l'),
    PrimitiveType.Int64:    lambda h: util.read_fmt(h, '<q'),
    PrimitiveType.SByte:    lambda h: util.read_fmt(h, 'c'),
    PrimitiveType.Single:   lambda h: util.read_fmt(h, 'f'),
    PrimitiveType.TimeSpan: _read_timespan,
    PrimitiveType.UInt16:   lambda h: util.read_fmt(h, '<H'),
    PrimitiveType.UInt32:   lambda h: util.read_fmt(h, '<L'),
    PrimitiveType.UInt64:   lambda h: util.read_fmt(h, '<Q'),
    PrimitiveType.Null:     lambda h: None,
    PrimitiveType.String:   _read_string,
}

def read_primitive(handle, primitive_type):
    try:
        return _PRIMITIVE_FACTORY_MAP[primitive_type](handle)
    except KeyError:
        raise NotImplementedError(
            'Reading primitive type {0} is not supported'.format(primitive_type))
